Level 1

AndreySapo opened the pull request.

AndreySapo committed:
добавлены инфрастуктурные файлы

 .gitignore, .editorconfig, .eslintrc.

AndreySapo committed:
инициализирован package.json

AndreySapo committed:
установлен eslint+airbnb

AndreySapo committed:
тестовый запуск сервера

AndreySapo committed:
создана схема user

AndreySapo committed:
создана схема movie

AndreySapo committed:
сделаны роуты и контроллеры

доработать обработку ошибок

AndreySapo committed:
добавлена обработка ошибок users

AndreySapo committed:
добавлена обработка ошибок movies

AndreySapo committed:
добавлен cors

AndreySapo committed:
добавлен cookieParser

AndreySapo committed:
добавлен логгер

AndreySapo committed:
cookies+dotenv

AndreySapo committed:
.env+удаление ненужных комментариев

AndreySapo committed:
очистка cookie по адресу signout

AndreySapo committed:
Update README.md

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [ ] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
  **Присутствуют блокировки вида `eslint-disable`.**
- [ ] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
**Ошибок не должно быть без блокировок.**
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [ ] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [ ] `GET /movies` — все сохранённые пользователем фильмы;
  **Запрос возвращает все фильмы в базе, а не фильмы текущего пользователя.**
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [ ] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
**Не защищен роут `'Этот путь не реализован'`.**
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [ ] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [ ] сообщение об ошибке соответсвует ее типу;
	**1. Не обработана ошибка дубля email при редактировании пользователя.**        
![2023-04-15_14-28-54](https://user-images.githubusercontent.com/86304001/232200825-3a2909e5-5eff-45ac-81b6-05618a26c5b1.png)
        **2. Неверное сообщение об ошибке. `CastError` выбрасывается при невалидном `_id`. Да, невалидного `_id` нет в базе, но смысл ошибки именно в невалидности переданных данных.**        
![2023-04-15_14-32-27](https://user-images.githubusercontent.com/86304001/232201024-093ed8a9-34f3-42b7-a8cf-3c8627e9f1b2.png)
        **3. Неверные сообщения об ошибке.**        
![2023-04-15_14-34-19](https://user-images.githubusercontent.com/86304001/232201076-05c98230-0e18-402e-a129-ac3d8cfc0257.png)
![2023-04-15_14-35-12](https://user-images.githubusercontent.com/86304001/232201107-94bb3e59-93b7-4f41-bdff-9ca9a712278d.png)
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  **1. Лишнее поле `owner` в запросе создания фильма.**
  **2. Лишнее ограничение длины пароля (нет в ТЗ).**
  - [ ] Данные валидируются перед добавлением в базу. 
  **1. Поле `movieId` на уровне схемы фильма должно валидироваться как число.**
  **2. Поле `name` на уровне схемы пользователя должно быть обязательным и не должно иметь значения по умолчанию.**
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [ ] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
**Можно обратиться только по **http** протоколу.**
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [ ] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
**Роуты подключены отдельно.**
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--2.14--> Валидация описана в отдельном модуле.
**Валидация описана в роутах.**
- [ ] <!--2.14--> Настроено логгирование:
  - [ ] запросы и ответы записываются в файл request.log;
  **Логгер запросов подключен после лимитера. Запросы, превышающие лимит, не попадают в лог.**
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [ ] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
**Обработчик описан в `app.js`.**
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [ ] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  **Адрес прописан в `app.js`.**
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Сообщения прописаны вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [ ] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
**Модуль сконфигурирован в `app.js`.**
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 39

## Доп. Комментарии:
1. **Лишняя проверка. Роут защищен авторизацией.**
![2023-04-15_14-36-48](https://user-images.githubusercontent.com/86304001/232201170-bd7dc935-adb0-4a7e-918c-4fb6b34a8f62.png)

AndreySapo committed:
review1

1. /signout не стал трогать, у меня он работает.
2. https буду делать уже на этапе женитьбы фронта и бэка
3. сообщения будут прописаны вручную, т.к. так проще находить ошибки

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [ ] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [ ] сообщение об ошибке соответсвует ее типу;
        **Неверное сообщение об ошибке.**               
![2023-04-16_02-01-02](https://user-images.githubusercontent.com/86304001/232250878-2295bc68-7a57-41fd-bbee-d4a0cd17bb29.png)
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [ ] <!--7.28--> Валидация реализована верно:
  - [ ] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  **Лишнее ограничение длины пароля (нет в ТЗ).**
  - [x] Данные валидируются перед добавлением в базу. 
- [ ] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
**Адрес общий для всех режимов.**
- [ ] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
**Можно обратиться только по **http** протоколу.**
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--2.14--> Валидация описана в отдельном модуле.
**Валидация для запроса редактирования пользователя описана в роуте.**
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Сообщения прописаны вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [ ] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
**Модуль сконфигурирован в `app.js`.**
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 71

AndreySapo committed:
review2

StarchenkovYaroslav commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id. Директивы eslint-disable, eslint-disable-line и eslint-disable-next-line запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды npx eslint . они должны отсутствовать.
- [x] <!--3.64--> В разделе scripts файла package.json есть команды:
  - [x] npm run start, которая запускает сервер на localhost:3000;
  - [x] npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно:
  - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
  - [x] `PATCH /users/me` — обновляет информацию о пользователе;
  - [x] `GET /movies` — все сохранённые пользователем фильмы;
  - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
  - [x] `DELETE /movies/movieId` — удаляет сохранённый фильмы по `_id`;
  - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
  - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
  - [ ] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [ ] <!--7.28--> Ошибки API обрабатываются:
  - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
  - [ ] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
	- [ ] сообщение об ошибке соответсвует ее типу;
        **Неверное сообщение об ошибке. Запрос отвечает за обновление пользователя, а не за создание.**                      
![2023-04-16_21-57-19](https://user-images.githubusercontent.com/86304001/232325012-4613f21f-7653-4921-a753-cc6a270d40a0.png)
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в виде хеша;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
  - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации. 
  - [x] Данные валидируются перед добавлением в базу. 
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из process.env.
- [ ] <!--3.64--> К серверу можно обратиться по https по адресу домена, указанному в README.md. 
**Можно обратиться только по **http** протоколу.**
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] сообщения ответов и ошибок (константы приложения) вынесены в отдельный файл с константами.
  **Сообщения прописаны вручную.**
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co.
- [x] <!--0.71--> Для имени базы данных используется `bitfilmsdb`.

## Количество баллов: 88
